### DISCUSSION TITLE
How to use npm pkg components like react-country-region-selector with react-hook-form

### DISCUSSION THREAD

**Discussion by freedomfflow** [react-hook-form/react-hook-form#7385] [Q&A] [answered]
Using React, Mui5, & v7 of react-hook-form

I'm using "react-country-region-selector" npm package for a country & region drop down selection.  I'm unable to figure out how to use them within React-hook-form.
How can I wrap 3rd party components like this into the <Control /> component imported from react-hook-form?
I'm fairly new to React and new to react-hook-form, so if there is an example I can mimic that you can point me to that would be wonderful.

**Comment by bluebill1049** [comment]
can you share a codesandbox? this should be able to solve with `Controller` by attach onChange, onBlur and value

**Reply by freedomfflow** [reply]
https://34yzo.csb.app/
I have never used code sandbox, and this doesn't function -- it is the code as it is in my component right now, and this works, within my app, but the country & region are not part of the 'form data'... They need to use a Controller -- But I'm not sure how to make it work.  The onChange for each is a function I wrote, which simply sets a local state variable -- I played around with the controller syntax, but didn't figure it out as I don't really understand how it all works.  Thanks if you can assist.  I know this sandbox is somewhat useless as it doesn't function, but it will likely take me all day to figure out how to make it work within sandbox, so I figured I'd spend that time doing more googling and trial and error working on this issue directly in my app :-)

**Reply by freedomfflow** [reply]
This doesn't work -- I've tried to modify this and the error I get typically is 'onChange is not a function'...

`            <Controller
              render={({ name, onChange, value}) =>
                  <CountryDropdown
                      defaultOptionLabel="Select Country"
                      name={name}
                      value={value}
                      onChange={onChange}
                  />
              }
              name='rcrs-country'
              value={country}
              onChange={selectCountry}
              control={control}
              defaultValue=''
            />`

**Reply by freedomfflow** [reply]
I believe I got it -- I made progress anyways... I need to use the 'field' syntax 

render={({ field: {name, onChange, value} })

**Reply by freedomfflow** [reply]
Well, I was wrong - this doesn't work, so if you have time perhaps you can assist?
I have to use 'useWatch' I believe, as the useState() is problematic to manage my onChange functions... either the onChange doesn't work or if it does, the value of country is not in my form data -- based on whether I pass it as a prop or not

Anyway, is there an example of how to use useWatch that involves a Controller as well to render my components?

I need to render <CountrySelector /> and have the result of that selection be available to <RegionSelector /> but I don't understand how to do this and I can't find any examples to copy.

Thanks

**Reply by freedomfflow** [reply]
You will laugh at me as I now think I have it working with useWatch --- I was WAAAAY over complicating things.... I started fresh, got rid of my useState, go rid of my event handlers, and simply wrapped the components in <Controller with just control and field, and it all works -- Everything I was doing was just interferring with how it works out of the box...

**Reply by freedomfflow** [reply]
and useWatch gives me the dependency I need between country and state -- so I don't have to manage anything -- just set up use watch and wrapped components in controller and I'm done -- 10 hours later -- haha

**Reply by bluebill1049** [reply]
I will take a look at today for you.

**Reply by bluebill1049** [reply]
![Screen Shot 2021-12-30 at 9 48 34 am](https://user-images.githubusercontent.com/10513364/147708875-e9515414-84f3-430e-88aa-78ee4d9f0e52.png)
looks like you forgot to save

**Comment by bluebill1049** [comment, marked as answer]
https://codesandbox.io/s/adoring-forest-5lbmr?file=/src/App.js

```tsx
import { CountryDropdown } from "react-country-region-selector";
import { useForm, Controller } from "react-hook-form";

import "./styles.css";

export default function App() {
  const { control, handleSubmit } = useForm({
    defaultValues: {
      test: ""
    }
  });

  return (
    <form
      onSubmit={handleSubmit((data) => {
        console.log(data);
      })}
    >
      <Controller
        name="test"
        render={({ field: { name, onChange, value } }) => (
          <CountryDropdown
            defaultOptionLabel="Select Country"
            name={name}
            value={value}
            onChange={onChange}
          />
        )}
        control={control}
      />
      <button>submit</button>
    </form>
  );
}
```
